core/kernel/user_ta.c

```diff
@@ -49,12 +49,26 @@
 #include <utee_defines.h>
 #include <util.h>
 
-static void init_utee_param(struct utee_params *up,
+static TEE_Result init_utee_param(struct utee_params *up,
 			const struct tee_ta_param *p, void *va[TEE_NUM_PARAMS])
 {
+	TEE_Result res = TEE_SUCCESS;
 	size_t n;
 
-	up->types = p->types;
+	/*
+	 * Note: The type of up->types is uint64_t, while the type of p->types
+	 * is uint32_t. To avoid potential uninitialized memory issue, first
+	 * clear 8 bytes (`&up->types`) first, and fill-in 4 bytes copied from
+	 * `p->types` there.
+	 */
+	res = clear_user(&up->types, sizeof(up->types));
```

Comment:
How about initializing a `struct utee_params` bounce buffer instead? That should allow us to use a single call to `copy_to_user()` at the end of this function instead.

Reply:
That's a good idea. Applied